idseq_dag/steps/run_validate_input.py

```diff
@@ -0,0 +1,212 @@
+
+        file_ext = self.additional_attributes.get("file_ext")
+        assert file_ext == 'fastq' or file_ext == 'fasta', 'Invalid file extension'
+
+        for i in range(num_inputs):
+            file = input_files[i]
+
+            # unzip if .gz file
+            if file[-3:] == '.gz':
+                cmd = f"gunzip {file}"
+                command.execute(cmd)
+                input_files[i] = file = file[:-3]
+
+        # keep a dictionary of the distribution of read lengths in the files
+        self.summary_dict = {'<50':0, '50-500': 0, '500-10000': 0, '10000+': 0}
+
+        try:
+            quick_check_passed = \
+                self.quick_check_file(input_files[0], file_ext == 'fastq') and \
+                (num_inputs == 1 or self.quick_check_file(input_files[1], file_ext == 'fastq'))
+
+            for infile, outfile in zip(input_files, output_files):
+                if quick_check_passed:
+                    self.truncate_file(infile, outfile, file_ext == 'fastq', max_fragments)
+                else:
+                    self.full_check_and_truncate_file(file, outfile, file_ext == 'fastq', max_fragments)
+
+            with open(summary_file, 'w') as summary_f:
+                json.dump(self.summary_dict, summary_f)
+        except Exception as e:
+            with open(summary_file, 'w') as summary_f:
+                json.dump({'Validation error': str(e)}, summary_f)
+            raise e
+
+        return
+
+    # quick_check_file returns:
+    #   True if the first 100 fragments all have the same length of reads and
+    #   are well-formated single-line FASTA / FASTQ entries. 
+    #
+    #   False if the entries are not formatted simply or the read lengths are not
+    #   all identical or if there is another possibly recoverable error
+    #
+    #   Throws an exception in the case of an unrecoverable abnormality
+    def quick_check_file(self, file, is_fastq, max_fragments_to_check = 100):
+        num_fragments = 0
+        fragment_length = 0
+
+        with open(file, 'r', encoding='utf-8') as input_f:
+            while True:
+                num_fragments += 1
+                if num_fragments > max_fragments_to_check:
+                    break
+
+                identifier_l = input_f.readline()
+                if len(identifier_l) == 0: # EOF
+                    break
+
+                read_l = input_f.readline()
+                if len(read_l) == 0: # unexpected EOF
+                    raise RuntimeError(f"Invalid input file, unexpected EOF: {file}")
+
+                if is_fastq:
+                    identifier2_l = input_f.readline()
+                    if len(identifier2_l) == 0:
+                        raise RuntimeError(f"Invalid FASTQ file, unexpected EOF: {file}")
+
+                    quality_l = input_f.readline()
+                    if len(quality_l) == 0:
+                        raise RuntimeError(f"Invalid FASTQ file, unexpected EOF: {file}")
```

Comment:
should check that len(quality_l ) == len(read_l)?

Reply:
we actually don't want to abort because of read_len != quality_len so we either truncate quality or pad it to match read_len